File: ios/Classes/Loader/PowerImageLoader.m
//
//  PowerImageLoader.m
//  power_image
//
//  Created by 王振辉 on 2021/7/14.
//

#import "PowerImageLoader.h"
#import "PowerImageRequestConfig.h"


@interface PowerImageLoader ()
@property (nonatomic, strong) NSMutableDictionary *imageLoaders;
@end

@implementation PowerImageLoader
+ (instancetype)sharedInstance {
    static dispatch_once_t onceToken;
    static PowerImageLoader *instance;
    dispatch_once(&onceToken, ^{
        instance = [[PowerImageLoader alloc] init];
    });
    return instance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _imageLoaders = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (void)registerImageLoader:(id<PowerImageLoaderProtocol>)request forType:(NSString *)type {
    self.imageLoaders[type] = request;
}

- (void)handleRequest:(PowerImageRequestConfig *)requestConfig completed:(PowerImageLoaderCompletionBlock)completedBlock {
    id <PowerImageLoaderProtocol>handler = self.imageLoaders[requestConfig.imageType];
    [handler handleRequest:requestConfig completed:completedBlock];
}
@end

File: android/src/main/java/com/taobao/power_image/loader/PowerImageLoader.java
package com.taobao.power_image.loader;

import com.taobao.power_image.request.PowerImageRequestConfig;

import java.util.HashMap;
import java.util.Map;

/**
 * created by Muke on 2021/7/22
 */
public class PowerImageLoader implements PowerImageLoaderProtocol {

    private final Map<String, PowerImageLoaderProtocol> imageLoaders;

    private PowerImageLoader() {
        imageLoaders = new HashMap<>();
    }

    private static class Holder {
        private final static PowerImageLoader instance = new PowerImageLoader();
    }

    public static PowerImageLoader getInstance() {
        return PowerImageLoader.Holder.instance;
    }

    public void registerImageLoader(PowerImageLoaderProtocol loader, String imageType) {
        imageLoaders.put(imageType, loader);
    }

    @Override
    public void handleRequest(PowerImageRequestConfig request, PowerImageResponse response) {
        PowerImageLoaderProtocol imageLoader = imageLoaders.get(request.imageType);
        if (imageLoader == null) {
            throw new IllegalStateException("PowerImageLoader for "
                    + request.imageType + " has not been registered.");
        }
        imageLoader.handleRequest(request, response);
    }
}
